README.md
node-red-contrib-keeper
=======================

<a href="http://nodered.org" target="_new">Node-RED</a> node to save and
retrieve document from Nunux Keeper v2.

Please note that this node is not compatible with Nunux Keeper v1.

![Screenshot](screenshot.png)

Install
-------

Run the following command in the root directory of your Node-RED install

    npm install node-red-node-keeper

Usage
-----

### Request node

Create, update, delete or retrieve a document from Nunux Keeper v2.

The document on Keeper is taken from the node `docid` property or the
`msg.docid` property.

The document content is sent as `msg.payload` property. If the request fails
`msg.error` will contain an error object.

To create a document The message payload should be as follow:

- `title`: Document title
- `label`: Label to classify the document (The label ID)
- `content`: Document content
- `contentType`: Document type
- `origin`: Document origin URL

All attributes are optional. If only `origin` attribute is provided, then the
document will be extracted from the endpoint.


CHANGELOG.md
Changelog
=========

0.0.5
-----

### Features

* Update Nunux Keeper client version

0.0.3
-----

### Features

* Using official Nunux Keeper client (node-keeper)

### Fix

* Put error into the message payload

0.0.2
-----

### Features

* Refreshing permanently the credentials with the refresh token
* Extract display name from the token

0.0.1
-----

### Features

* Credentials
* Create/Update/Delete/Retrieve a document



keeper-client.js
'use strict'

const request = require('request')
const url = require('url')

function base64urlUnescape (str) {
  str += new Array(5 - str.length % 4).join('=')
  return str.replace(/\-/g, '+').replace(/_/g, '/')
}

function base64urlDecode (str) {
  return new Buffer(base64urlUnescape(str), 'base64').toString()
}

class KeeperApi {
  constructor (client) {
    this.client = client
  }

  postDocument (doc) {
    return this.client._request({
      method: 'POST',
      url: this.client.credentials.apiSite + '/v2/document',
      json: true,
      body: doc
    })
  }

  getDocument (docid) {
    return this.client._request({
      method: 'GET',
      url: this.client.credentials.apiSite + '/v2/document/' + docid,
      json: true
    })
  }

  updateDocument (doc, update) {
    return this.client._request({
      method: 'PUT',
      url: this.client.credentials.apiSite + '/v2/document/' + doc.id,
      json: true,
      body: update
    })
  }

  removeDocument (doc) {
    return this.client._request({
      method: 'DELETE',
      url: this.client.credentials.apiSite + '/v2/document/' + doc.id,
      json: true
    })
  }
}

class KeeperClient {
  constructor (credentials, options, onUpdateCredentials) {
    // TODO Add debug flag
    this.credentials = Object.assign({
      authSite: 'http://login.nunux.org',
      authPath: '/auth/realms/NunuxKeeper/protocol/openid-connect/auth',
      tokenPath: '/auth/realms/NunuxKeeper/protocol/openid-connect/token',
      apiSite: 'http://api.nunux.org/keeper'
    }, credentials)
    this.options = Object.assign({
      debug: false
    }, options)
    this.onUpdateCredentials = onUpdateCredentials || function (cred) {
      this._log('Credential updated but no callback registered.')
    }.bind(this)
    this.api = new KeeperApi(this)
  }

  _log () {
    const p1 = arguments[0]
    const pn = Array.prototype.slice.call(arguments, 1)
    console.log.apply(console, ['KeeperClient: ' + p1].concat(pn))
  }

  _debug () {
    if (this.options.debug) {
      this._log.apply(this, arguments)
    }
  }

  _error () {
    const p1 = arguments[0]
    const pn = Array.prototype.slice.call(arguments, 1)
    console.error.apply(console, ['KeeperClient: ' + p1].concat(pn))
  }

  static decodeAccessToken (token) {
    const segments = token.split('.')
    if (segments.length !== 3) {
      return {email: 'Unknown'}
    }
    const payload = JSON.parse(base64urlDecode(segments[1]))
    return payload
  }

  authorizeURL (redirect_uri, state) {
    const u = url.parse(this.credentials.authSite)
    return url.format({
      protocol: u.protocol,
      hostname: u.hostname,
      pathname: this.credentials.authPath,
      query: {
        response_type: 'code',
        client_id: this.credentials.clientId,
        state: state,
        redirect_uri: redirect_uri
      }
    })
  }

  _tokenURL () {
    const u = url.parse(this.credentials.authSite)
    return url.format({
      protocol: u.protocol,
      hostname: u.hostname,
      pathname: this.credentials.tokenPath
    })
  }

  token (redirect_uri, code) {
    this._debug('Getting token...', code)
    return new Promise((resolve, reject) => {
      request.post({
        url: this._tokenURL(),
        json: true,
        form: {
          grant_type: 'authorization_code',
          code: code,
          client_id: this.credentials.clientId,
          client_secret: this.credentials.clientSecret,
          redirect_uri: redirect_uri
        }
      }, (err, result, data) => {
        if (err) {
          this._error('Unable to get token (http error).', err)
          return reject(err)
        }
        if (data.error) {
          this._error('Unable to get token.', data.error)
          return reject(data.error)
        }
        this.credentials.accessToken = data.access_token
        this.credentials.refreshToken = data.refresh_token
        this.credentials.expiresIn = data.expires_in
        this.credentials.expireTime = data.expires_in + (new Date().getTime() / 1000)
        this.credentials.tokenType = data.token_type
        const decoded = KeeperClient.decodeAccessToken(this.credentials.accessToken)
        this.credentials.displayName = decoded.email || decoded.name
        // Notify new credentials
        this.onUpdateCredentials(this.credentials)
        return resolve(this.credentials)
      })
    })
  }

  enableAutoRefreshToken () {
    if (this.refreshTimeout) {
      clearTimeout(this.refreshTimeout)
    }
    // Compute the refresh tiemout one minute before expiration
    const timeout = this.credentials.expireTime - Date.now() - 60000
    this._debug('Expiration time:', this.credentials.expireTime)
    this._debug('Computed timeout:', timeout)
    this.refreshTimeout = setTimeout(() => {
      this._refreshToken()
    }, timeout)
  }

  _refreshToken () {
    if (!this.credentials.refreshToken) {
      this._error('Unable to refresh. No refresh token.')
      return Promise.reject('ENOREFRESHTOKEN')
    }
    return new Promise((resolve, reject) => {
      this._debug('Refreshing the token...')
      request.post({
        url: this._tokenURL(),
        json: true,
        form: {
          grant_type: 'refresh_token',
          client_id: this.credentials.clientId,
          client_secret: this.credentials.clientSecret,
          refresh_token: this.credentials.refreshToken
        }
      }, (err, result, data) => {
        if (err) {
          this._error('Unable to get refresh token (http error).', err)
          return reject(err)
        }
        if (data.error) {
          this._error('Unable to get refresh token.', data)
          return reject(data)
        }
        this.credentials.accessToken = data.access_token
        this.credentials.expiresIn = data.expires_in
        this.credentials.expireTime = Date.now() + (data.expires_in * 1000)
        this.credentials.tokenType = data.token_type
        if (data.refresh_token) {
          this.credentials.refreshToken = data.refresh_token
          if (this.refreshTimeout) {
            this.enableAutoRefreshToken()
          }
        }
        // Notify new credentials
        this.onUpdateCredentials(this.credentials)
        return resolve(this.credentials)
      })
    })
  }

  _request (req, retries) {
    retries = retries || 1
    if (typeof req !== 'object') {
      req = { url: req }
    }
    req.method = req.method || 'GET'
    this._debug('%s %s', req.method, req.url)
    if (!req.hasOwnProperty('json')) {
      req.json = true
    }
    // Set access token
    req.auth = { bearer: this.credentials.accessToken }

    // Trigger refresh token if access token is expired
    if (!this.credentials.expireTime || this.credentials.expireTime < (new Date().getTime() / 1000)) {
      if (retries === 0) {
        this._error('Too many refresh attempts.')
        return Promise.reject('ETOOMANYREFRESH')
      }
      return this._refreshToken()
      .then(() => this._request(req, 0))
    }

    return new Promise((resolve, reject) => {
      request(req, (err, result, data) => {
        if (err) {
          this._error('Request error (http error).', err)
          return reject(err)
        }
        if (data.error) {
          this._error('Request error.', data.error)
          return reject(data.error)
        }
        if (result.statusCode === 401 && retries > 0) {
          retries = retries - 1
          this._error('401 received. Trying to refresh the token...')
          return this._refreshToken()
          .then(() => this._request(req, retries))
          .then(resolve)
          .catch(reject)
        }
        if (result.statusCode >= 400) {
          return reject(data)
        }
        return resolve(data)
      })
    })
  }
}

module.exports = KeeperClient



keeper.js
/**
 * Copyright 2016 Nunux Org.
 *
 * Licensed under the Apache License, Version 2.0 (the 'License')
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

const DEBUG = process.env.NODE_ENV === 'development'

module.exports = function (RED) {
  'use strict'

  const KeeperSDK = require('node-keeper')
  const crypto = require('crypto')
  const sdk = new KeeperSDK({debug: DEBUG})

  function KeeperNode (n) {
    RED.nodes.createNode(this, n)
  }

  RED.nodes.registerType('keeper-credentials', KeeperNode, {
    credentials: {
      displayName:  {type: 'text'},
      clientId:     {type: 'text'},
      clientSecret: {type: 'password'},
      accessToken:  {type: 'password'},
      refreshToken: {type: 'password'},
      expireTime:   {type: 'password'}
    }
  })

  RED.httpAdmin.get('/keeper-credentials/auth', function (req, res) {
    if (!req.query.clientId || !req.query.clientSecret || !req.query.id || !req.query.callback) {
      res.send(400)
      return
    }
    const node_id = req.query.id
    const callback = req.query.callback
    const options = {
      clientId: req.query.clientId,
      clientSecret: req.query.clientSecret
    }
    sdk.createClient(options)
    .then((client) => {
      const csrfToken = crypto.randomBytes(18).toString('base64').replace(/\//g, '-').replace(/\+/g, '_')
      const credentials = Object.assign({
        csrfToken: csrfToken,
        callback: callback
      }, options)
      res.cookie('csrf', csrfToken)
      res.redirect(client.getAuthorizeURL(callback, node_id + ':' + csrfToken))
      RED.nodes.addCredentials(node_id, credentials)
    })
  })

  RED.httpAdmin.get('/keeper-credentials/auth/callback', function (req, res) {
    if (req.query.error) {
      return res.send('ERROR: ' + req.query.error + ': ' + req.query.error_description)
    }
    const state = req.query.state.split(':')
    const node_id = state[0]
    const credentials = RED.nodes.getCredentials(node_id)
    if (!credentials || !credentials.clientId || !credentials.clientSecret) {
      return res.send(RED._('keeper.error.no-credentials'))
    }
    // console.log('Credentials:' + JSON.stringify(credentials))
    if (state[1] !== credentials.csrfToken) {
      return res.status(401).send(
        RED._('keeper.error.token-mismatch')
      )
    }

    sdk.createClient(credentials)
    .then((client) => {
      client.getTokens(credentials.callback, req.query.code)
      .then((creds) => {
        RED.nodes.addCredentials(node_id, Object.assign({}, credentials, creds))
        return res.send(RED._('keeper.error.authorized'))
      })
      .catch((err) => {
        console.log('Keeper client error: ' + err)
        return res.send(RED._('keeper.error.something-broke'))
      })
    })
  })

  function KeeperRequestNode (n) {
    RED.nodes.createNode(this, n)
    this.action = n.action || null
    this.docid = n.docid || ''
    this.keeper = RED.nodes.getNode(n.keeper)
    var node = this
    if (!this.keeper || !this.keeper.credentials.accessToken) {
      this.warn(RED._('keeper.warn.missing-credentials'))
      return
    }

    sdk.createClient(Object.assign(this.keeper.credentials, {
      onRefreshTokens: (creds) => {
        RED.nodes.addCredentials(node.id, Object.assign({}, this.keeper.credentials, creds))
      }
    }), this.keeper.credentials)
    .then((client) => {
      node.on('input', function (msg) {
        const action = node.action || msg.action
        const docid = node.docid || msg.docid || msg.payload.id
        if (action !== null && action !== 'post') {
          if (!docid) {
            node.error(RED._('keeper.error.no-docid-specified'))
            return
          }
        }
        let result
        switch (action) {
          case 'get':
            result = client.api.document.get(docid)
            break
          case 'post':
            result = client.api.document.post(msg.payload)
            break
          case 'put':
            result = client.api.document.update({id: docid}, msg.payload)
            break
          case 'delete':
            result = client.api.document.remove({id: docid})
            break
          default:
            node.error(RED._('keeper.error.no-action-specified'))
            return
        }
        node.status({fill: 'blue', shape: 'dot', text: 'keeper.status.requesting'})
        result.then((data) => {
          msg.payload = data
          delete msg.error
          node.status({})
          node.send(msg)
        }).catch((err) => {
          node.error(RED._('keeper.error.request-failed', {err: JSON.stringify(err)}))
          node.status({fill: 'red', shape: 'ring', text: 'keeper.status.failed'})
          msg.payload = err
          msg.error = err
          msg.statusCode = 502
          node.send(msg)
        })
      })
    })
  }
  RED.nodes.registerType('keeper', KeeperRequestNode)
}
